src/Lean/Data/Lsp/Basic.lean

```diff
@@ -0,0 +1,323 @@
+/-
+Copyright (c) 2020 Marc Huisinga. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+
+Authors: Marc Huisinga, Wojciech Nawrocki
+-/
+import Lean.Data.Json
+
+/-! Defines most of the 'Basic Structures' in the LSP specification
+(https://microsoft.github.io/language-server-protocol/specifications/specification-current/),
+as well as some utilities.
+
+Since LSP is Json-based, Ints/Nats are represented by Floats on the wire. -/
+
+namespace Lean
+namespace Lsp
+
+open Json
+
+abbrev DocumentUri := String
+
+/-- An array of the lines in a text document.
+Elements mustn't contain newline characters. -/
+def DocumentText := Array String
+
+/-- Can represent both a UTF-16 position - this is what LSP clients send -
+and a codepoint position - what we use internally.
+`character` is accepted liberally: actual character := min(line length, character) -/
+structure Position := (line : Nat) (character : Nat)
+
+instance Position.hasFromJson : HasFromJson Position :=
+⟨fun j => do
+  line ← j.getObjValAs? Nat "line";
+  character ← j.getObjValAs? Nat "character";
+  pure ⟨line, character⟩⟩
+
+instance Position.hasToJson : HasToJson Position :=
+⟨fun o => mkObj $
+  [ ⟨"line", o.line⟩
+  , ⟨"character", o.character⟩
```

Comment:
style: We don't use leading commas. Here I would probably use 
```
⟨fun o => mkObj [
  ⟨"line", o.line⟩,
  ⟨"character", o.character⟩]⟩
```
Note: We do already support extraneous trailing commas in structure notation, but not yet in lists.

Reply:
We need a `leanfmt` :smile: